docs/opup.rst

```diff
@@ -0,0 +1,85 @@
+.. _opup:
+
+Budget Increase Utility
+========================
+
+Some opcode budget is consumed during execution of every Algorand Smart Contract because every teal
+instruction has a corresponding cost. In order for the evaluation to succeed, the budget consumed must not
+exceed the budget provided. This constraint may introduce a problem for expensive contracts that quickly
+consume the initial budget of 700. The OpUp budget increase utility provides a workaround using inner
+transactions. Note that there is a context specific limit to the number of inner transactions issued in a
+transaction group so budget cannot be increased arbitrarily. The available budget when using the OpUp
+utility will need to be high enough to execute the teal code that issues the inner transactions. A budget
+of ~20 is enough for most use cases.
+
+Usage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :any:`pyteal.OpUp` utility is available in two modes: :any:`Explicit` and :any:`OnCall`:
+
+================= ================================================================================
+OpUp Mode         Description
+================= ================================================================================
+:code:`Explicit`  Calls a user provided external app when more budget is requested.
+:code:`OnCall`    Creates the app to be called when more budget is requested.
+================= ================================================================================
+
+
+:any:`Explicit` has the benefit of constructing more lightweight inner transactions but requires the
+target app ID to be provided in the foreign apps array field of the transaction and the :any:`pyteal.OpUp`
+constructor in order for it to be accessible. :any:`OnCall` is easier to use but has slightly more overhead
+because the target app must be created and deleted during the evaluation of an app call.
+
+:any:`pyteal.OpUp.ensure_budget` attempts to ensure that the available budget is at least the budget requested by
+the caller. If there aren't enough funds for issuing the inner transactions or the inner transaction limit
+is exceeded, the evaluation will fail. The typical usage pattern is to insert the :any:`pyteal.OpUp.ensure_budget`
+call just before a particularly heavyweight subroutine or expression. Keep in mind that the required budget
+expression will be evaluated before the inner transactions are issued so it may be prudent to avoid expensive
+expressions, which may exhaust the budget before it can be increased.
+
+In the example below, the Ed25519Verify expression is used, which costs 1900.
```

Comment:
@algoidurovic Minor thoughts:
* It's possible to link the `Ed25519Verify` reference.   https://stackoverflow.com/questions/22700606/how-would-i-cross-reference-a-function-generated-by-autodoc-in-sphinx helped me understand how to do it.
* Optionally, I prefer to add commas for readability on numbers with > 3 digits.


```suggestion
In the example below, the :py:meth:`pyteal.Ed25519Verify` expression is used, which costs 1,900.
```